File: client/src/lib/models/workout.ts
import type { Category } from "./category";

export class Workout {
    id: string;
    user_id: string;
    name: string;
    category: Category;
    note: string;
}

export class WorkoutRow {
    id: string;
    user_id: string;
    name: string;
    category: Category;
    note: string;
}

export class InsertWorkoutRow {
    user_id: string;
    name: string;
    category: Category;
    note: string;
}

export class UpsertWorkoutRow<T> {
    workout_row: T;
    selected_routine_ids: string[];
    unselected_routine_ids: string[];
}

export class WorkoutHistoryRow {
    id: string;
    workout: string;
    user_id: string;
    completed_on: string;
    sets: Set[];
}

export class WorkoutHistory {
    id: string;
    workout: Workout;
    user_id: string;
    completed_on: string;
    sets: Set[];
}

export class Set {
    weight: number;
    reps: number;
}

File: server/src/model/workout.rs
use super::shared_types::{UserRef, WorkoutHistoryRef, WorkoutRef, Set};
use serde::{Deserialize, Serialize};


#[derive(Debug, Deserialize, Serialize)]
pub struct Workout {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<WorkoutRef>,
    pub user_id: UserRef,
    pub name: String,
    pub category: String,
    pub note: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct WorkoutHistoryRow {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<WorkoutHistoryRef>,
    pub user_id: UserRef,
    pub workout_id: WorkoutRef,
    pub name: String,
    pub completed_on: String,
    pub sets: Vec<Set>,
}